### PR TITLE
Containerize MongoDB and Redis for Kaze

### LINKED ISSUE #194: Containerize Redis and MongoDB databases required by kaze

### PR DESCRIPTION
**Pull request by aryaman-titan**
@alphadose Plz check the **docker-compose.yml** file and tell me if it's okay to continue with this for further using it in Kaze.

### CODE REVIEW

**Comment by supra08** [comment]
Well, as discussed over slack, we are not going with this approach (using docker-compose or dockerfiles). You might want to look how mongoDB service is actually setup in gasper and how you can apply the same for setting up gasper's mongoDB. Refer this:
https://sdslabs.slack.com/archives/CCAFCAWT0/p1586298950144200

**Comment by aryaman-titan** [comment]
> Well, as discussed over slack, we are not going with this approach (using docker-compose or dockerfiles). You might want to look how mongoDB service is actually setup in gasper and how you can apply the same for setting up gasper's mongoDB. Refer this:
> https://sdslabs.slack.com/archives/CCAFCAWT0/p1586298950144200

Okay cool! I'll just look upon the setting up of the service! :+1:

**Comment by karan0299** [comment]
Squash the last two commits

File: lib/database/handler.go [hunk 1]
```diff
@@ -47,6 +49,39 @@ func SetupDBInstance(databaseType string) (string, types.ResponseError) {
 				storedir,
 				env)
 		}
+	case types.MongoDBGasper:
+		{
+			dockerImage := configs.ImageConfig.Mongodb
+			port := strconv.Itoa(configs.ServiceConfig.Kaze.MongoDB.ContainerPort)
+			fmt.Println(port)
```
Comment (karan0299) [inline lib/database/handler.go]: Remove this

File: lib/database/handler.go [hunk 2]
```diff
@@ -47,6 +49,39 @@ func SetupDBInstance(databaseType string) (string, types.ResponseError) {
...
+			env := configs.ServiceConfig.Kaze.MongoDB.Env
+			workdir := "/data/db"
+			storedir := filepath.Join(storepath, "mongodb-storage")
```
Comment (karan0299) [inline lib/database/handler.go]:
```suggestion
			storedir := filepath.Join(storepath, "gasper-mongodb-storage")
```

File: lib/database/handler.go [hunk 3]
```diff
@@ -47,6 +49,39 @@ func SetupDBInstance(databaseType string) (string, types.ResponseError) {
...
+				databaseType)
+		}
+	case types.RedisGasper:
+		{
+			dockerImage := configs.ImageConfig.Redis
+			port := strconv.Itoa(configs.ServiceConfig.Kaze.Redis.ContainerPort)
+			fmt.Println(port)
+			env := configs.ServiceConfig.Kaze.Redis.Env
+			workdir := "/data/db"
+			storedir := filepath.Join(storepath, "redis-storage")
```
Comment (karan0299) [inline lib/database/handler.go]:
```suggestion
			storedir := filepath.Join(storepath, "gasper-redis-storage")
```

File: lib/database/handler.go [hunk 4]
```diff
@@ -47,6 +49,39 @@ func SetupDBInstance(databaseType string) (string, types.ResponseError) {
...
+				workdir,
+				storedir,
+				env,
+				databaseType)
+		}
+	case types.RedisGasper:
+		{
+			dockerImage := configs.ImageConfig.Redis
+			port := strconv.Itoa(configs.ServiceConfig.Kaze.Redis.ContainerPort)
+			fmt.Println(port)
```
Comment (karan0299) [inline lib/database/handler.go]: Remove print statement

**Review by karan0299 — CHANGES_REQUESTED** [review]
I have left some comments, I will test this branch by tomorrow and will let you know and any further changes are required

File: config.sample.toml [hunk 5]
```diff
@@ -108,6 +109,7 @@ golang = "sdsws/golang:1.1"
 ruby = "sdsws/ruby:1.0"
 mysql = "mysql:5.7"
 mongodb = "mongo:4.2.1"
+redis = "redis:latest"
```
Comment (alphadose) [inline config.sample.toml]: replace this with `redis:6.0-rc3-alpine3.11`

File: config.sample.toml [hunk 6]
```diff
@@ -54,6 +54,7 @@ password = "alphadose"
 url = "mongodb://localhost:27017"
 
 
+
```
Comment (alphadose) [inline config.sample.toml:56]: Remove this unnecessary line

File: config.sample.toml [hunk 7]
```diff
@@ -131,7 +133,15 @@ exposure_interval = 30
 # to all worker nodes and removes inactive nodes from the central registry-server.
 cleanup_interval = 600
 deploy = true   # Deploy Kaze?
-port = 3000
```
Comment (alphadose) [inline config.sample.toml:136]: why delete this ?

File: config.sample.toml [hunk 8]
```diff
@@ -131,7 +133,15 @@ exposure_interval = 30
...
+
+[services.kaze.mongodb]
+plugin = true  # Deploy MongoDB server and let `Kaze` manage it?
+container_port = 27019  # Port on which the MongoDB server container will run
+
+[services.kaze.redis]
+plugin = true  # Deploy Redis server and let `Kaze` manage it?
+container_port = 6380  # Port on which the Redis server container will run
```
Comment (alphadose) [inline config.sample.toml]: add environment variables for the user and password of both of them. See how it is done below for mongodb. Find the equivalent for redis

File: config.sample.toml [hunk 9]
```diff
@@ -54,6 +54,7 @@ password = "alphadose"
 url = "mongodb://localhost:27017"
```
Comment (alphadose) [inline config.sample.toml]: change it to the port on which your mongodb container is running

File: config.sample.toml [hunk 10]
```diff
@@ -54,6 +54,7 @@ password = "alphadose"
...
 
 
+
 ###########################
 #   Redis Configuration   #
 ###########################
```
Comment (alphadose) [inline config.sample.toml]: change the port here as well to the redis container you created

**Review by alphadose — CHANGES_REQUESTED** [review]

File: config.sample.toml [hunk 11]
```diff
@@ -51,7 +51,7 @@ password = "alphadose"
 [mongo]
 # For databases with authentication
 # use the following URL format `mongodb://username:password@host:port`.
-url = "mongodb://localhost:27017"
+url = "mongodb://localhost:27019"
```
Comment (alphadose) [inline config.sample.toml]: If you have populated these values
```
MONGO_INITDB_ROOT_USERNAME = "YOUR_ROOT_NAME"   # Root user of MongoDB server inside the container
MONGO_INITDB_ROOT_PASSWORD = "YOUR_ROOT_PASSWORD"   # Root password of MongoDB server inside the container
```

then the connection URL will be `mongodb://YOUR_ROOT_NAME:YOUR_ROOT_PASSWORD@localhost:27019`

**Review by alphadose — CHANGES_REQUESTED** [review]

File: config.sample.toml [hunk 12]
```diff
@@ -133,6 +134,18 @@ cleanup_interval = 600
...
+container_port = 27019  # Port on which the MongoDB server container will run
+
+# Environment variables for MongoDB docker container.
+[services.kaze.mongodb.env]
+MONGO_INITDB_ROOT_USERNAME = "alphadose"   # Root user of MongoDB server inside the container
+MONGO_INITDB_ROOT_PASSWORD = "alphadose"   # Root password of MongoDB server inside the container
+
+[services.kaze.redis]
+plugin = true  # Deploy Redis server and let `Kaze` manage it?
+container_port = 6380  # Port on which the Redis server container will run
```
Comment (alphadose) [inline config.sample.toml:153]: The password for redis container should be declared here and used when creating the container

File: lib/database/mongo.go [hunk 13]
```diff
@@ -16,6 +16,10 @@ var (
 	mongoRootUser     = configs.ServiceConfig.Kaen.MongoDB.Env["MONGO_INITDB_ROOT_USERNAME"]
 	mongoRootPassword = configs.ServiceConfig.Kaen.MongoDB.Env["MONGO_INITDB_ROOT_PASSWORD"]
 	mongoPort         = configs.ServiceConfig.Kaen.MongoDB.ContainerPort
+
+	mongoGasperUser     = configs.ServiceConfig.Kaze.MongoDB.Env["MONGO_INITDB_ROOT_USERNAME"]
+	mongoGasperPassword = configs.ServiceConfig.Kaze.MongoDB.Env["MONGO_INITDB_ROOT_PASSWORD"]
+	mongoGasperPort     = configs.ServiceConfig.Kaze.MongoDB.ContainerPort
```
Comment (alphadose) [inline lib/database/mongo.go]: Are these 3 needed ?

**Review by alphadose — CHANGES_REQUESTED** [review]

**Review by supra08 — APPROVED** [review]

File: lib/database/handler.go [hunk 14]
```diff
@@ -11,60 +10,76 @@ import (
...
+		env:         configs.ServiceConfig.Kaze.MongoDB.Env,
+		workdir:     "/data/db",
+		storedir:    filepath.Join(storepath, "gasper-mongodb-storage"),
+		containerID: docker.CreateMongoDBContainer,
+	},
+	types.MySQL: {
+		dockerImage: configs.ImageConfig.Mysql,
+		port:        strconv.Itoa(configs.ServiceConfig.Kaen.MySQL.ContainerPort),
+		env:         configs.ServiceConfig.Kaen.MySQL.Env,
+		workdir:     "/app",
```
Comment (alphadose) [inline lib/database/handler.go]: workdir should be `/var/lib/mysql`

`/app` will be the workdir for the new image not this one

**Review by alphadose — CHANGES_REQUESTED** [review]

**Review by alphadose — APPROVED** [review]

**Pull request merged**